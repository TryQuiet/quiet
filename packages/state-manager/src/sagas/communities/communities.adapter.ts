import { type Community } from '@quiet/types'
import { createEntityAdapter } from '@reduxjs/toolkit'

export const communitiesAdapter = createEntityAdapter<Community>({
    selectId: community => community.id,
})
