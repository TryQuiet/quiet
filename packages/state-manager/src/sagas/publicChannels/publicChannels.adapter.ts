import { ChannelMessage, PublicChannelStatus, PublicChannelStorage, PublicChannelSubscription } from '@quiet/types'
import { createEntityAdapter } from '@reduxjs/toolkit'

export const publicChannelsAdapter = createEntityAdapter<PublicChannelStorage>({
  selectId: (channel) => channel.id
})

export const publicChannelsStatusAdapter = createEntityAdapter<PublicChannelStatus>({
  selectId: (channel) => channel.id
})

export const publicChannelsSubscriptionsAdapter = createEntityAdapter<PublicChannelSubscription>({
  selectId: (channel) => channel.id
})

export const channelMessagesAdapter = createEntityAdapter<ChannelMessage>()
