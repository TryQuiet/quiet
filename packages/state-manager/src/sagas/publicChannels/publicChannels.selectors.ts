import { createSelector } from 'reselect'
import { StoreKeys } from '../store.keys'
import {
  publicChannelsAdapter,
  channelMessagesAdapter,
  publicChannelsStatusAdapter,
  publicChannelsSubscriptionsAdapter,
} from './publicChannels.adapter'
import { type CreatedSelectors, type StoreState } from '../store.types'
import { allUsers } from '../users/users.selectors'
import { userProfiles } from '../users/userProfile/userProfile.selectors'
import { formatMessageDisplayDay } from '../../utils/functions/dates/formatMessageDisplayDate'
import { displayableMessage } from '../../utils/functions/dates/formatDisplayableMessage'
import { isDefined } from '@quiet/common'
import {
  type ChannelMessage,
  type DisplayableMessage,
  MessageType,
  type MessagesDailyGroups,
  type MessagesGroupsType,
  type PublicChannel,
  type PublicChannelStatus,
  INITIAL_CURRENT_CHANNEL_ID,
  UserProfile,
} from '@quiet/types'

const selectState: CreatedSelectors[StoreKeys.PublicChannels] = (state: StoreState) => state[StoreKeys.PublicChannels]

export const selectChannels = createSelector(selectState, state => {
  if (!state) return []
  return publicChannelsAdapter.getSelectors().selectAll(state.channels)
})

const selectChannelsSubscriptions = createSelector(selectState, state => {
  if (!state) return []
  return publicChannelsSubscriptionsAdapter.getSelectors().selectAll(state.channelsSubscriptions)
})

const pendingGeneralChannelRecreation = createSelector(selectState, state => {
  return state.pendingGeneralChannelRecreation
})

export const subscribedChannels = createSelector(selectChannelsSubscriptions, subscriptions => {
  return subscriptions.map(subscription => {
    if (subscription.subscribed) return subscription.id
  })
})

// Serves for testing purposes only
export const selectGeneralChannel = createSelector(selectChannels, channels => {
  const draft = channels.find(item => item.name === 'general')
  if (!draft) {
    console.error('No general channel')
    return
  }
  const channel: PublicChannel = {
    name: draft.name,
    description: draft.description,
    owner: draft.owner,
    timestamp: draft.timestamp,
    id: draft.id,
  }
  return channel
})

export const publicChannels = createSelector(selectChannels, selectChannelsSelector => {
  const channels = Array.from(selectChannelsSelector)
  const sorted = channels.sort((a, b) => {
    if (a.name === 'general') {
      return -1
    }
    if (b.name === 'general') {
      return 0
    }
    return a.name.localeCompare(b.name)
  })

  return sorted
})

export const sortedChannels = createSelector(publicChannels, channels => {
  const sorted = channels.sort((a, b) => {
    if (a.name === 'general') {
      return -1
    }
    if (b.name === 'general') {
      return 0
    }
    return a.name.localeCompare(b.name)
  })

  return sorted
})

export const generalChannel = createSelector(publicChannels, publicChannelsSelector => {
  return publicChannelsSelector.find(channel => channel.name === 'general')
})

export const currentChannelId = createSelector(selectState, generalChannel, (state, general) => {
  if (!state) {
    return undefined
  }
  if (state.currentChannelId === INITIAL_CURRENT_CHANNEL_ID) {
    if (general) {
      return general.id
    } else {
      return undefined
    }
  } else {
    return state.currentChannelId
  }
})

export const recentChannels = createSelector(
  publicChannels,
  generalChannel,
  (publicChannelsSelector, generalChannelSelector) => {
    const recentChannels = publicChannelsSelector.sort((a, b) => b.timestamp - a.timestamp).slice(0, 3)
    return recentChannels.length >= 3 ? recentChannels : [generalChannelSelector]
  }
)

export const getChannelById = (channelId: string) =>
  createSelector(publicChannels, channels => {
    const channel = channels.find(channel => channel.id === channelId)
    if (!channel) {
      console.log('channel dont exist')
    }
    return channel
  })

export const dynamicSearchedChannels = (channelInput: string) =>
  createSelector(publicChannels, recentChannels, (publicChannelsSelector, recentChannelsSelector) => {
    const filteredList = publicChannelsSelector.filter(channel => channel.name.includes(channelInput))

    const isFilteredList = filteredList.length > 0 ? filteredList : recentChannelsSelector

    const channelList = channelInput.length === 0 ? recentChannelsSelector : isFilteredList

    return channelList.filter(isDefined)
  })

// Is being used in tests
export const currentChannel = createSelector(currentChannelId, selectChannels, (id, channels) => {
  if (!id) return undefined
  return channels.find(channel => channel.id === id)
})

export const currentChannelName = createSelector(currentChannel, channel => {
  if (!channel) return ''
  return channel.name
})

export const currentChannelMessages = createSelector(currentChannel, channel => {
  if (!channel) return []
  return channelMessagesAdapter.getSelectors().selectAll(channel.messages)
})

export const sortedCurrentChannelMessages = createSelector(currentChannelMessages, messages => {
  return messages.sort((a, b) => b.createdAt - a.createdAt).reverse()
})

export const currentChannelLastDisplayedMessage = createSelector(sortedCurrentChannelMessages, messages => {
  return messages[0]
})

export const newestCurrentChannelMessage = createSelector(sortedCurrentChannelMessages, messages => {
  return messages[messages.length - 1]
})

export const displayableCurrentChannelMessages = createSelector(
  sortedCurrentChannelMessages,
  allUsers,
  (messages, users) => {
    return messages.reduce((result: DisplayableMessage[], message: ChannelMessage) => {
      const user = users[message.pubKey]
      if (user) {
        // @ts-ignore
        result.push(displayableMessage(message, user))
      }
      return result
    }, [])
  }
)

export const currentChannelMessagesCount = createSelector(displayableCurrentChannelMessages, messages => {
  return messages.length
})

/**
 * Channel messages grouped by day
 */
export const dailyGroupedCurrentChannelMessages = createSelector(displayableCurrentChannelMessages, messages => {
  const result: MessagesGroupsType = messages.reduce((groups: MessagesGroupsType, message: DisplayableMessage) => {
    const date = formatMessageDisplayDay(message.date)

    if (!groups[date]) {
      groups[date] = []
    }

    groups[date].push(message)
    return groups
  }, {})

  return result
})

/**
 * Channel messages grouped by day and then additionally by sender (if
 * there are successive messages by the same sender). Also adds a
 * profile photo to the first message of each user.
 */
export const currentChannelMessagesMergedBySender = createSelector(
  dailyGroupedCurrentChannelMessages,
  userProfiles,
  (groups: MessagesGroupsType, userProfiles: Record<string, UserProfile>) => {
    const result: MessagesDailyGroups = {}
    for (const day in groups) {
      result[day] = groups[day].reduce((merged: DisplayableMessage[][], message: DisplayableMessage) => {
        if (!merged.length) {
          // Add profile photo to user's first message
          if (message.pubKey in userProfiles) {
            message.photo = userProfiles[message.pubKey].profile.photo
          }
          merged.push([message])
          return merged
        }

        // Get last item from collected array for comparison
        const index = merged.length && merged.length - 1
        const last = merged[index][0]

        if (
          last?.pubKey === message?.pubKey &&
          message.createdAt - last.createdAt < 300 &&
          message.type !== MessageType.Info &&
          last.type !== MessageType.Info
        ) {
          merged[index].push(message)
        } else {
          // Add profile photo to user's first message
          if (message.pubKey in userProfiles) {
            message.photo = userProfiles[message.pubKey].profile.photo
          }
          merged.push([message])
        }

        return merged
      }, [])
    }

    return result
  }
)

export const channelsStatus = createSelector(selectState, state => {
  if (!state?.channelsStatus) return {}
  return publicChannelsStatusAdapter.getSelectors().selectEntities(state.channelsStatus)
})

export const channelsStatusSorted = createSelector(selectState, state => {
  if (!state?.channelsStatus) return []
  const statuses = publicChannelsStatusAdapter.getSelectors().selectAll(state.channelsStatus)

  return statuses
    .sort((a, b) => {
      const aCreatedAt = a.newestMessage?.createdAt
      const bCreatedAt = b.newestMessage?.createdAt
      if (!aCreatedAt && !bCreatedAt) return 0
      if (!aCreatedAt) return -1
      if (!bCreatedAt) return 1
      return aCreatedAt - bCreatedAt
    })
    .reverse()
})

export const unreadChannels = createSelector(channelsStatus, status => {
  return Object.values(status)
    .filter(isDefined)
    .reduce((result: string[], channel: PublicChannelStatus) => {
      if (channel.unread) {
        result.push(channel.id)
      }
      return result
    }, [])
})

export const publicChannelsSelectors = {
  publicChannels,
  subscribedChannels,
  currentChannelId,
  currentChannelName,
  currentChannel,
  currentChannelMessages,
  sortedCurrentChannelMessages,
  displayableCurrentChannelMessages,
  newestCurrentChannelMessage,
  currentChannelMessagesCount,
  currentChannelMessagesMergedBySender,
  currentChannelLastDisplayedMessage,
  unreadChannels,
  channelsStatus,
  channelsStatusSorted,
  dynamicSearchedChannels,
  sortedChannels,
  pendingGeneralChannelRecreation,
  generalChannel,
  getChannelById,
}
