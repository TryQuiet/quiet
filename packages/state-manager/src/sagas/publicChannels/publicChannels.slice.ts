import { createSlice, EntityState, PayloadAction } from '@reduxjs/toolkit'
import { StoreKeys } from '../store.keys'
import {
  publicChannelsAdapter,
  channelMessagesAdapter,
  publicChannelsStatusAdapter,
  publicChannelsSubscriptionsAdapter
} from './publicChannels.adapter'

import logger from '../../utils/logger'
import {
  CacheMessagesPayload,
  ChannelDeletionResponsePayload,
  ChannelMessage,
  ChannelsReplicatedPayload,
  ClearMessagesCachePayload,
  CreateChannelPayload,
  CreatedChannelResponse,
  DeleteChannelFromStorePayload,
  DeleteChannelPayload,
  DisableChannelPayload,
  Identity,
  INITIAL_CURRENT_CHANNEL_ID,
  MarkUnreadChannelPayload,
  PublicChannelStatus,
  PublicChannelStorage,
  PublicChannelSubscription,
  SendInitialChannelMessagePayload,
  SendNewUserInfoMessagePayload,
  SetChannelSubscribedPayload,
  SetCurrentChannelPayload,
  UpdateNewestMessagePayload
} from '@quiet/types'
const log = logger('publicChannels')

export class PublicChannelsState {
  public currentChannelId: string = INITIAL_CURRENT_CHANNEL_ID

  public pendingGeneralChannelRecreation: boolean = false

  public channels: EntityState<PublicChannelStorage> = publicChannelsAdapter.getInitialState()

  public channelsStatus: EntityState<PublicChannelStatus> =
    publicChannelsStatusAdapter.getInitialState()

  public channelsSubscriptions: EntityState<PublicChannelSubscription> =
    publicChannelsSubscriptionsAdapter.getInitialState()
}

export const publicChannelsSlice = createSlice({
  initialState: { ...new PublicChannelsState() },
  name: StoreKeys.PublicChannels,
  reducers: {
    createChannel: (state, _action: PayloadAction<CreateChannelPayload>) => state,
    deleteChannel: (state, _action: PayloadAction<DeleteChannelPayload>) => state,
    channelDeletionResponse: (state, _action: PayloadAction<ChannelDeletionResponsePayload>) =>
      state,
    deleteChannelFromStore: (state, action: PayloadAction<DeleteChannelFromStorePayload>) => {
      const { channelId } = action.payload

      publicChannelsSubscriptionsAdapter.removeOne(state.channelsSubscriptions, channelId)
      publicChannelsStatusAdapter.removeOne(state.channelsStatus, channelId)
      publicChannelsAdapter.removeOne(state.channels, channelId)
    },
    disableChannel: (state, action: PayloadAction<DisableChannelPayload>) => {
      const { channelId } = action.payload
      publicChannelsAdapter.updateOne(state.channels, {
        id: channelId,
        changes: {
          disabled: true
        }
      })
    },
    clearMessagesCache: (state, action: PayloadAction<ClearMessagesCachePayload>) => {
      const { channelId } = action.payload
      const channel = state.channels.entities[channelId]
      if (!channel) return
      channelMessagesAdapter.setAll(channel.messages, [])
    },
    startGeneralRecreation: state => {
      state.pendingGeneralChannelRecreation = true
    },
    finishGeneralRecreation: state => {
      state.pendingGeneralChannelRecreation = false
    },
    createGeneralChannel: state => state,
    sendInitialChannelMessage: (state, _action: PayloadAction<SendInitialChannelMessagePayload>) =>
      state,
    sendNewUserInfoMessage: (state, _action: PayloadAction<SendNewUserInfoMessagePayload>) => state,
    addChannel: (state, action: PayloadAction<CreatedChannelResponse>) => {
      const { channel } = action.payload
      publicChannelsAdapter.addOne(state.channels, {
        ...channel,
        messages: channelMessagesAdapter.getInitialState()
      })
      publicChannelsStatusAdapter.addOne(state.channelsStatus, {
        id: channel.id,
        unread: false,
        newestMessage: null
      })
    },
    setChannelSubscribed: (state, action: PayloadAction<SetChannelSubscribedPayload>) => {
      const { channelId } = action.payload
      publicChannelsSubscriptionsAdapter.upsertOne(state.channelsSubscriptions, {
        id: channelId,
        subscribed: true
      })
    },
    channelsReplicated: (state, _action: PayloadAction<ChannelsReplicatedPayload>) => state,
    setCurrentChannel: (state, action: PayloadAction<SetCurrentChannelPayload>) => {
      const { channelId } = action.payload
      state.currentChannelId = channelId
    },
    cacheMessages: (state, action: PayloadAction<CacheMessagesPayload>) => {
      const { messages, channelId } = action.payload
      const channel = state.channels.entities[channelId]
      if (!channel) return
      channelMessagesAdapter.setAll(channel.messages, messages)
    },

    markUnreadChannel: (state, action: PayloadAction<MarkUnreadChannelPayload>) => {
      const { channelId } = action.payload
      publicChannelsStatusAdapter.updateOne(state.channelsStatus, {
        id: channelId,
        changes: {
          unread: true
        }
      })
    },
    clearUnreadChannel: (state, action: PayloadAction<MarkUnreadChannelPayload>) => {
      const { channelId } = action.payload
      publicChannelsStatusAdapter.updateOne(state.channelsStatus, {
        id: channelId,
        changes: {
          unread: false
        }
      })
    },
    updateNewestMessage: (state, action: PayloadAction<UpdateNewestMessagePayload>) => {
      const { message } = action.payload
      publicChannelsStatusAdapter.updateOne(state.channelsStatus, {
        id: message.channelId,
        changes: {
          newestMessage: message
        }
      })
    },
    // Utility action for testing purposes
    test_message: (
      state,
      action: PayloadAction<{
        // [x: string]: ChannelMessage
        message: ChannelMessage
        identity: Identity
        verifyAutomatically: boolean
      }>
    ) => {
      const { message } = action.payload
      const channel = state.channels.entities[message.channelId]
      if (!channel) return
      channelMessagesAdapter.addOne(channel.messages, message)
    }
  }
})

export const publicChannelsActions = publicChannelsSlice.actions
export const publicChannelsReducer = publicChannelsSlice.reducer
