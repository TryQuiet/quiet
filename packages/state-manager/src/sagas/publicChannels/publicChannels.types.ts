import { DisplayableMessage } from '@quiet/types'
import { type Dictionary, type EntityState } from '@reduxjs/toolkit'
import { type FileMetadata } from '../files/files.types'

export const INITIAL_CURRENT_CHANNEL_ID = 'initialcurrentChannelId'

export interface PublicChannel {
  name: string
  description: string
  owner: string
  timestamp: number
  id: string
  disabled?: boolean
}

export interface PublicChannelStorage extends PublicChannel {
  messages: EntityState<ChannelMessage>
}

export interface PublicChannelStatus {
  id: string
  unread: boolean
  newestMessage: ChannelMessage | null
}

export interface PublicChannelSubscription {
  id: string
  subscribed: boolean
}

export interface ChannelMessage {
  id: string
  type: number
  message: string
  createdAt: number
  channelId: string
  signature: string
  pubKey: string
  media?: FileMetadata
}

export type MessagesDailyGroups = Record<string, DisplayableMessage[][]>

export interface ChannelsReplicatedPayload {
  channels: Dictionary<PublicChannel>
}

export interface CreateChannelPayload {
  channel: PublicChannel
}

export interface DeleteChannelPayload {
  channelId: string
}

export interface DeleteChannelResponse {
  channelId: string
}

export interface CreatedChannelResponse {
  channel: PublicChannel
}

export interface SetChannelSubscribedPayload {
  channelId: string
}

export interface SetCurrentChannelPayload {
  channelId: string
}

export interface SetChannelMessagesSliceValuePayload {
  messagesSlice: number
  channelId: string
}

export interface PendingMessage {
  message: ChannelMessage
}

export interface SendInitialChannelMessagePayload {
  channelName: string
  channelId: string
}
export interface SendNewUserInfoMessagePayload {
  certificates: string[]
}

export interface IncomingMessages {
  messages: ChannelMessage[]
  isVerified?: boolean
}

export interface CacheMessagesPayload {
  messages: ChannelMessage[]
  channelId: string
}

export interface MarkUnreadChannelPayload {
  channelId: string
  message?: ChannelMessage
}

export interface UpdateNewestMessagePayload {
  message: ChannelMessage
}

export interface DeleteChannelFromStorePayload {
  channelId: string
}

export interface ClearMessagesCachePayload {
  channelId: string
}

export interface DisableChannelPayload {
  channelId: string
}

export function instanceOfChannelMessage(object: any): object is ChannelMessage {
  return 'channelId' in object
}
