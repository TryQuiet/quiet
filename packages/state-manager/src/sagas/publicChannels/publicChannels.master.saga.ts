import { type Socket } from '../../types'
import { all, takeEvery } from 'typed-redux-saga'
import { publicChannelsActions } from './publicChannels.slice'
import { createChannelSaga } from './createChannel/createChannel.saga'
import { deleteChannelSaga } from './deleteChannel/deleteChannel.saga'
import { createGeneralChannelSaga } from './createGeneralChannel/createGeneralChannel.saga'
import { sendInitialChannelMessageSaga } from './createGeneralChannel/sendInitialChannelMessage.saga'
import { sendNewUserInfoMessageSaga } from './sendNewUserInfoMessage/sendNewUserInfoMessage.saga'
import { clearUnreadChannelsSaga } from './markUnreadChannels/markUnreadChannels.saga'
import { channelsReplicatedSaga } from './channelsReplicated/channelsReplicated.saga'
import { channelDeletionResponseSaga } from './channelDeletionResponse/channelDeletionResponse.saga'

export function* publicChannelsMasterSaga(socket: Socket): Generator {
    yield all([
        takeEvery(publicChannelsActions.createChannel.type, createChannelSaga, socket),
        takeEvery(publicChannelsActions.deleteChannel.type, deleteChannelSaga, socket),
        takeEvery(publicChannelsActions.channelDeletionResponse.type, channelDeletionResponseSaga),
        takeEvery(publicChannelsActions.createGeneralChannel.type, createGeneralChannelSaga),
        takeEvery(publicChannelsActions.sendInitialChannelMessage.type, sendInitialChannelMessageSaga),
        takeEvery(publicChannelsActions.channelsReplicated.type, channelsReplicatedSaga),
        takeEvery(publicChannelsActions.setCurrentChannel.type, clearUnreadChannelsSaga),
        takeEvery(publicChannelsActions.sendNewUserInfoMessage.type, sendNewUserInfoMessageSaga),
    ])
}
