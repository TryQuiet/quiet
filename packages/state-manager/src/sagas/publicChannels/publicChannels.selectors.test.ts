import { setupCrypto } from '@quiet/identity'
import { Store } from '../store.types'
import { FactoryGirl } from 'factory-girl'
import { getFactory, publicChannels } from '../..'
import { prepareStore } from '../../utils/tests/prepareStore'
import {
  publicChannels as getPublicChannels,
  currentChannelMessagesMergedBySender,
  sortedCurrentChannelMessages,
  displayableCurrentChannelMessages,
  publicChannelsSelectors
} from './publicChannels.selectors'
import { publicChannelsActions } from './publicChannels.slice'

import { identityActions } from '../identity/identity.slice'
import { usersActions } from '../users/users.slice'
import {
  formatMessageDisplayDate,
  formatMessageDisplayDay
} from '../../utils/functions/dates/formatMessageDisplayDate'
import { displayableMessage } from '../../utils/functions/dates/formatDisplayableMessage'
import { DateTime } from 'luxon'
import { generateChannelId } from '@quiet/common'
import {
  ChannelMessage,
  Community,
  DisplayableMessage,
  Identity,
  MessageType,
  PublicChannel
} from '@quiet/types'
import { communitiesActions } from '../communities/communities.slice'

describe('publicChannelsSelectors', () => {
  let store: Store
  let factory: FactoryGirl

  let community: Community
  let alice: Identity
  let john: Identity

  let generalChannel: PublicChannel
  let channelIdes: string[] = []

  const msgs: { [id: string]: ChannelMessage } = {}
  const msgsOwners: { [id: string]: string } = {}

  beforeAll(async () => {
    setupCrypto()

    // Set date display format
    process.env.LC_ALL = 'en_US.UTF-8'

    store = prepareStore().store

    factory = await getFactory(store)

    community = await factory.create<
      ReturnType<typeof communitiesActions.addNewCommunity>['payload']
    >('Community')

    alice = await factory.create<ReturnType<typeof identityActions.addNewIdentity>['payload']>(
      'Identity',
      { id: community.id, nickname: 'alice' }
    )
    const generalChannelState = publicChannelsSelectors.generalChannel(store.getState())
    if (generalChannelState) generalChannel = generalChannelState
    expect(generalChannel).not.toBeUndefined()
    channelIdes = [...channelIdes, generalChannel.id]
    john = await factory.create<ReturnType<typeof identityActions.addNewIdentity>['payload']>(
      'Identity',
      { id: community.id, nickname: 'john' }
    )
    store.dispatch(publicChannelsActions.setCurrentChannel({ channelId: generalChannel.id }))
    // Setup channels
    const channelNames = ['croatia', 'allergies', 'sailing', 'pets', 'antiques']

    for (const name of channelNames) {
      const channel = await factory.create<
        ReturnType<typeof publicChannels.actions.addChannel>['payload']
      >('PublicChannel', {
        channel: {
          name: name,
          description: `Welcome to #${name}`,
          timestamp: DateTime.utc().valueOf(),
          owner: alice.nickname,
          id: generateChannelId(name)
        }
      })
      channelIdes = [...channelIdes, channel.channel.id]
    }

    /* Messages ids are being used only for veryfing proper order...
    ...they have no impact on selectors work */
    const messages = [
      {
        id: '1',
        createdAt: DateTime.fromObject({
          year: 2020,
          month: 10,
          day: 20,
          hour: 5,
          minute: 50
        }).toSeconds(),
        identity: alice
      },
      // Message 2 and 3 have info type, so they are tested for not being grouped together.
      {
        id: '2',
        type: 3,
        createdAt: DateTime.fromObject({
          year: 2020,
          month: 10,
          day: 20,
          hour: 6,
          minute: 10
        }).toSeconds(),
        identity: alice
      },
      {
        id: '3',
        type: 3,
        createdAt: DateTime.fromObject({
          year: 2020,
          month: 10,
          day: 20,
          hour: 6,
          minute: 11,
          second: 30,
          millisecond: 1
        }).toSeconds(),
        identity: alice
      },
      {
        id: '4',
        createdAt: DateTime.fromObject({
          year: 2020,
          month: 10,
          day: 20,
          hour: 6,
          minute: 11,
          second: 30,
          millisecond: 2
        }).toSeconds(),
        identity: alice
      },
      {
        id: '5',
        createdAt: DateTime.fromObject({
          year: 2020,
          month: 10,
          day: 20,
          hour: 6,
          minute: 12,
          second: 1
        }).toSeconds(),
        identity: john
      },
      {
        id: '6',
        createdAt: DateTime.fromObject({
          year: 2020,
          month: 10,
          day: 20,
          hour: 6,
          minute: 12,
          second: 2
        }).toSeconds(),
        identity: alice
      },
      {
        id: '7',
        createdAt: DateTime.fromObject({
          year: 2021,
          month: 2,
          day: 5,
          hour: 18,
          minute: 2
        }).toSeconds(),
        identity: alice
      },
      {
        id: '8',
        createdAt: DateTime.fromObject({
          year: 2021,
          month: 2,
          day: 5,
          hour: 20,
          minute: 50
        }).toSeconds(),
        identity: alice
      },
      {
        id: '9',
        createdAt: DateTime.fromObject({
          year: DateTime.now().toUTC().year,
          month: DateTime.now().toUTC().month,
          day: DateTime.now().toUTC().day,
          hour: 20,
          minute: 50
        }).toSeconds(),
        identity: alice
      }
    ]

    // Shuffle messages array
    const shuffled = messages
      .map(value => ({ value, sort: Math.random() }))
      .sort((a, b) => a.sort - b.sort)
      .map(({ value }) => value)

    for (const item of shuffled) {
      const message = await factory.create<
        ReturnType<typeof publicChannelsActions.test_message>['payload']
      >('Message', {
        identity: item.identity,
        message: {
          id: item.id,
          type: item.type || MessageType.Basic,
          message: `message_${item.id}`,
          createdAt: item.createdAt,
          channelId: generalChannel.id,
          signature: '',
          pubKey: ''
        },
        verifyAutomatically: true
      })
      msgs[item.id] = message.message
      msgsOwners[item.id] = item.identity.nickname
    }
  })

  it('get messages sorted by date', async () => {
    const messages = sortedCurrentChannelMessages(store.getState())

    const formattedMessages = messages.reduce((prev: ChannelMessage[], curr: ChannelMessage) => {
      return [
        ...prev,
        {
          ...curr,
          channelId: 'general_ec4bca1fa76046c53dff1e49979c3647'
        }
      ]
    }, [])

    formattedMessages.forEach(message => {
      expect(message).toMatchSnapshot({
        createdAt: expect.any(Number),
        pubKey: expect.any(String),
        signature: expect.any(String)
      })
    })
  })

  it('get grouped messages', async () => {
    const messages = currentChannelMessagesMergedBySender(store.getState())
    // Convert regular messages to displayable messages
    const displayable: { [id: string]: DisplayableMessage } = {}
    for (const message of Object.values(msgs)) {
      displayable[message.id] = displayableMessage(message, msgsOwners[message.id])
    }

    // Get groups names
    const groupDay1 = formatMessageDisplayDay(formatMessageDisplayDate(msgs['7'].createdAt))
    expect(groupDay1).toBe('Feb 05')
    const groupDay2 = formatMessageDisplayDay(formatMessageDisplayDate(msgs['1'].createdAt))
    expect(groupDay2).toBe('Oct 20')
    const groupDay3 = formatMessageDisplayDay(formatMessageDisplayDate(msgs['9'].createdAt))
    expect(groupDay3).toBe('Today')

    const expectedGrouppedMessages = {
      [groupDay1]: [[displayable['7']], [displayable['8']]],
      [groupDay2]: [
        [displayable['1']],
        [displayable['2']],
        [displayable['3']],
        [displayable['4']],
        [displayable['5']],
        [displayable['6']]
      ],
      [groupDay3]: [[displayable['9']]]
    }
    expect(messages).toStrictEqual(expectedGrouppedMessages)
  })

  it('get channel list in a consistent order', async () => {
    const channels = getPublicChannels(store.getState()).map(channel => channel.name)

    expect(channels).toStrictEqual([
      'general',
      'allergies',
      'antiques',
      'croatia',
      'pets',
      'sailing'
    ])
  })

  it("don't select messages without author", async () => {
    const channelId = generateChannelId('utah')
    const channel = (
      await factory.create<ReturnType<typeof publicChannels.actions.addChannel>['payload']>(
        'PublicChannel',
        {
          channel: {
            name: 'utah',
            description: 'Welcome to #utah',
            timestamp: DateTime.utc().valueOf(),
            owner: alice.nickname,
            id: channelId
          }
        }
      )
    ).channel

    const elouise = await factory.create<
      ReturnType<typeof identityActions.addNewIdentity>['payload']
    >('Identity', { id: community.id, nickname: 'elouise' })

    if (!elouise.userCertificate) throw new Error('no elouise.userCertificate')
    store.dispatch(
      usersActions.test_remove_user_certificate({ certificate: elouise.userCertificate })
    )

    store.dispatch(
      publicChannelsActions.setCurrentChannel({
        channelId: channel.id
      })
    )

    await factory.create<ReturnType<typeof publicChannelsActions.test_message>['payload']>(
      'Message',
      {
        identity: elouise,
        message: {
          id: '0',
          type: MessageType.Basic,
          message: 'elouise_message',
          createdAt: DateTime.now().valueOf(),
          channelId: channel.id,
          signature: '',
          pubKey: ''
        },
        verifyAutomatically: true
      }
    )

    const messages = displayableCurrentChannelMessages(store.getState())

    expect(messages.length).toBe(0)
  })

  it('unreadChannels return empty object if PublicChannels is in the wrong state (no channelStatus)', async () => {
    // This case occurred in a built app
    const store = prepareStore().store
    const factory = await getFactory(store)
    await factory.create<ReturnType<typeof communitiesActions.addNewCommunity>['payload']>(
      'Community'
    )

    const oldState = store.getState()
    const channelId = oldState.PublicChannels.channels.ids[0]
    const newState = {
      ...oldState,
      PublicChannels: {
        ...oldState.PublicChannels,
        channels: {
          ids: oldState.PublicChannels.channels.ids,
          entities: {
            [channelId]: {
              ...oldState.PublicChannels.channels.entities[channelId],
              channelsStatus: undefined
            }
          }
        }
      }
    }
    // @ts-expect-error
    const unreadChannels = publicChannelsSelectors.unreadChannels(newState)
    expect(unreadChannels).toEqual([])
  })

  it('unreadChannels selector returns only unread channels', async () => {
    const channelId = channelIdes.find(channelId => channelId.includes('allergies'))
    if (!channelId) throw new Error('no channel id')
    store.dispatch(
      publicChannelsActions.markUnreadChannel({
        channelId
      })
    )
    const unreadChannels = publicChannelsSelectors.unreadChannels(store.getState())
    expect(unreadChannels).toEqual([channelId])
  })
})

export {}
