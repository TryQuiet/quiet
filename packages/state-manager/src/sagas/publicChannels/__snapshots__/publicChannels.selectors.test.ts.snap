// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publicChannelsSelectors get messages sorted by date 1`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "1",
  "message": "message_1",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 1,
}
`;

exports[`publicChannelsSelectors get messages sorted by date 2`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "2",
  "message": "message_2",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 3,
}
`;

exports[`publicChannelsSelectors get messages sorted by date 3`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "3",
  "message": "message_3",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 3,
}
`;

exports[`publicChannelsSelectors get messages sorted by date 4`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "4",
  "message": "message_4",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 1,
}
`;

exports[`publicChannelsSelectors get messages sorted by date 5`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "5",
  "message": "message_5",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 1,
}
`;

exports[`publicChannelsSelectors get messages sorted by date 6`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "6",
  "message": "message_6",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 1,
}
`;

exports[`publicChannelsSelectors get messages sorted by date 7`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "7",
  "message": "message_7",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 1,
}
`;

exports[`publicChannelsSelectors get messages sorted by date 8`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "8",
  "message": "message_8",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 1,
}
`;

exports[`publicChannelsSelectors get messages sorted by date 9`] = `
Object {
  "channelId": "general_ec4bca1fa76046c53dff1e49979c3647",
  "createdAt": Any<Number>,
  "id": "9",
  "message": "message_9",
  "pubKey": Any<String>,
  "signature": Any<String>,
  "type": 1,
}
`;
